src/asm3/animal.py

```diff
@@ -3847,8 +3851,169 @@ def update_diary_linkinfo(dbo: Database, animalid: int, a: ResultRow = None, dia
+                    if row not in validrows:
+                        validrows.append(row)
+    for row in animallocations.copy():
+        if row not in validrows:
+            dbo.delete("animallocation", row["ID"], username)
+            animallocations.remove(row)
+            rowid = row["ID"]
+            rowprevlocationid = row["PREVANIMALLOCATIONID"]
+            dbo.execute(
+                "UPDATE animallocation SET PrevAnimalLocationID = ? WHERE AnimalID = ? AND PrevAnimalLocationID = ?",
+                (rowprevlocationid, animalid, rowid)
+            )
+    
+    ## Detect and delete obsolete external movements
+    obsoletemovements = []
+    for row in animallocations:
+        if row["MOVEMENTID"]:
+            movementfound = False
+            for movement in movementrows:
+                if row["MOVEMENTID"] == movement["ID"]:
+                    movementfound = True
+                    break
+            if not movementfound:
+                obsoletemovements.append(row)
+    for row in obsoletemovements:
+        dbo.delete("animallocation", row["ID"], username)
+        animallocations.remove(row)
+        rowid = row["ID"]
+        rowprevlocationid = row["PREVANIMALLOCATIONID"]
+        dbo.execute(
+            "UPDATE animallocation SET PrevAnimalLocationID = ? WHERE AnimalID = ? AND PrevAnimalLocationID = ?",
+            (rowprevlocationid, animalid, rowid)
+        )
+
+    ## Sync external movements
+    for movementrow in movementrows:
+        outboundmovementfound = False
+        for locationrow in animallocations:
+            if locationrow["MOVEMENTID"] == movementrow["ID"]:
+                outboundmovementfound = True
+                break
+        if outboundmovementfound:
+            if asm3.i18n.remove_time(locationrow["DATE"]) != asm3.i18n.remove_time(movementrow["MOVEMENTDATE"]):
+                dbo.execute(
+                    "UPDATE animallocation SET Date = ? WHERE ID = ?",
+                    (movementrow["MOVEMENTDATE"], locationrow["ID"])
+                )
+        else:
+            fromid = 0
+            fromunit = ''
+            mostrecentmovement = False
+            for row in animallocations:
+                if asm3.i18n.remove_time(row["DATE"]) <= asm3.i18n.remove_time(movementrow["MOVEMENTDATE"]):
+                    mostrecentmovement = row
+                else:
+                    break
+            if mostrecentmovement:
+                fromid = mostrecentmovement["TOLOCATIONID"]
+                fromunit = mostrecentmovement["TOUNIT"]
+            insert_animallocation(dbo, username, animalid, animalname, sheltercode, fromid, fromunit, 0, '*', 0, movementrow["ID"])
+        
+        if movementrow["RETURNDATE"]:
+            idmatchfound = 0
+            for locationrow in animallocations:
+                if locationrow["MOVEMENTID"] == movementrow["ID"]:
+                    if idmatchfound == 0:
+                        idmatchfound = 1
```

Comment:
This idmatch thing is really confusing! I think it would be more straightforward if we look for the matching record, update it if we find and then create it if we didn't? Unless I missed something else this is doing

```
updated = False
locationrow = find_location(movementid=movementrow.ID)
if locationrow:
        ... update existing record
else:
    insert_animallocation ...
```    

see comment below for the find_location bit